interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,997 @@
+}
+
+// RuleWithID returns the rule with the given ID.
+// If the rule is not found, returns ErrRuleNotFound.
+// If the rule does not apply to the given user, returns ErrUserNotAllowed.
+func (rdb *RuleDB) RuleWithID(user uint32, id prompting.IDType) (*Rule, error) {
+	rdb.mutex.RLock()
+	defer rdb.mutex.RUnlock()
+	return rdb.ruleWithIDForUser(user, id)
+}
+
+// Creates a rule with the given information and adds it to the rule database.
+// If any of the given parameters are invalid, returns an error. Otherwise,
+// returns the newly-added rule, and saves the database to disk.
+func (rdb *RuleDB) AddRule(user uint32, snap string, iface string, constraints *prompting.Constraints, outcome prompting.OutcomeType, lifespan prompting.LifespanType, duration string) (*Rule, error) {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	newRule, err := rdb.populateNewRule(user, snap, iface, constraints, outcome, lifespan, duration)
+	if err != nil {
+		return nil, err
+	}
+	if err, conflicts, conflictingPermission := rdb.addRuleToTree(newRule); err != nil {
+		return nil, fmt.Errorf("%w: conflicts: %+v, Permission: '%s'", err, conflicts, conflictingPermission)
+	}
+	rdb.addRule(newRule)
+
+	if err := rdb.save(); err != nil {
+		rdb.removeExistingRuleNoError(newRule)
+		return nil, err
+	}
+
+	rdb.notifyRule(user, newRule.ID, nil)
+	return newRule, nil
+}
+
+// RemoveRule the rule with the given ID from the rules database. If the rule
+// does not apply to the given user, returns ErrUserNotAllowed. If successful,
+// saves the database to disk.
+func (rdb *RuleDB) RemoveRule(user uint32, id prompting.IDType) (*Rule, error) {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	rule, err := rdb.ruleWithIDForUser(user, id)
+	if err != nil {
+		return nil, err
+	}
+	err = rdb.removeRuleFromTree(rule)
+	// If error occurs, rule was still fully removed from tree
+	rdb.removeRuleWithID(id)
+
+	if err := rdb.save(); err != nil {
+		// TODO rollback the change to remove rule?
+		return nil, err
+	}
+
+	data := map[string]string{"removed": "removed"}
+	rdb.notifyRule(user, id, data)
+	return rule, err
+}
+
+// RemoveRulesForSnap removes all rules pertaining to the given snap for the
+// user with the given user ID.
+func (rdb *RuleDB) RemoveRulesForSnap(user uint32, snap string) []*Rule {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	ruleFilter := func(rule *Rule) bool {
+		return rule.User == user && rule.Snap == snap
+	}
+	rules := rdb.rulesInternal(ruleFilter)
+	rdb.removeRulesInternal(user, rules)
+	return rules
+}
+
+// removeRulesInternal removes all of the given rules from the rule DB and
+// records a notice for each one.
+//
+// The caller must ensure that the database lock is held for writing.
+func (rdb *RuleDB) removeRulesInternal(user uint32, rules []*Rule) error {
+	for _, rule := range rules {
+		rdb.removeExistingRuleNoError(rule)
+		data := map[string]string{"removed": "removed"}
+		rdb.notifyRule(user, rule.ID, data)
+	}
+	return rdb.save()
+}
```

Comment:
As above, I think we can be safer about this by rolling back the changes if the save fails:
```suggestion
func (rdb *RuleDB) removeRulesInternal(user uint32, rules []*Rule) error {
	for _, rule := range rules {
		// Remove rule from the rules list. Caller should ensure that
		// the rule exists, and thus this should not error.
		rdb.removeRuleWithID(rule.ID)
	}
	
	// Now that rules have been removed from rules list, attempt to save
	if err := rdb.save(); err != nil {
		// Roll back the change by re-adding all removed rules
		for _, rule := range rules {
			rdb.addRule(rule)
		}
		return err
	}
	
	// Save successful, now remove rules' variants from tree
	data := map[string]string{"removed": "removed"}
	for _, rule := range rules {
		rdb.removeRuleFromTree(rule)
		// If error occurs, rule was still fully removed from tree, and no other
		// rule was affected. We want the rule fully removed, so this is fine.
		rdb.notifyRule(user, rule.ID, data)
	}
	return nil
}
```